File: server/fastify/config/lifecycles/26_runtime_template_resolver.lifecycle.mjs
import fp from "fastify-plugin";
import { createResolver, MissingStrategy } from "@ployglot/runtime-template-resolver";

async function resolverPlugin(fastify, _opts) {
  if (!fastify.envResolve) {
    throw new Error(
      "envResolve missing — confirm 15_env_resolve.lifecycle.mjs is loaded before 26_runtime_template_resolver"
    );
  }
  if (!fastify.compute_registry) {
    throw new Error(
      "compute_registry missing — confirm 24_compute_functions.lifecycle.mjs runs before 26_runtime_template_resolver"
    );
  }
  // IGNORE so unmatched {{app.X}} / {{request.X}} refs (no context at boot) become
  // literal strings rather than aborting. {{fn:…}} refs resolve via the registry.
  const resolver = createResolver({
    registry: fastify.compute_registry,
    missingStrategy: MissingStrategy.IGNORE,
  });

  fastify.decorate("runtime_template_resolver", resolver);
  fastify.decorateRequest("runtime_template_resolver", null);
  fastify.addHook("onRequest", async (req) => {
    req.runtime_template_resolver = resolver;
  });
}

const wrapped = fp(resolverPlugin, {
  name: "runtime-template-resolver",
  fastify: ">=4",
  dependencies: ["env-resolve", "compute-functions"],
});

export default async function lifecycle(server, _config) {
  await server.register(wrapped);
  server.log.info("runtime-template-resolver decorator registered");
}

File: server/fastapi/config/lifecycles/26_runtime_template_resolver.lifecycle.py
"""Construct a ResolverHandle on app.state for Depends() injection."""

from __future__ import annotations

import logging

from ._runtime_template_resolver import ResolverHandle, make_resolver

log = logging.getLogger("fastapi_server.runtime_template_resolver")


def on_init(app, _config) -> None:
    app.state.runtime_template_resolver = None


async def on_startup(app, _config) -> None:
    env_resolve = getattr(app.state, "env_resolve", None)
    if env_resolve is None:
        raise RuntimeError(
            "env_resolve missing — confirm 15_env_resolve.lifecycle.py runs before 26_runtime_template_resolver"
        )
    registry = getattr(app.state, "compute_registry", None)
    if registry is None:
        raise RuntimeError(
            "compute_registry missing — confirm 24_compute_functions.lifecycle.py runs before 26"
        )
    app.state.runtime_template_resolver = ResolverHandle(
        env_resolve=env_resolve,
        _resolver=make_resolver(registry=registry),
    )
    log.info("runtime_template_resolver initialized (registry has %d fns)", len(getattr(app.state, "compute_registry_ledger", {}).get("registered", [])))


async def on_shutdown(app, _config) -> None:
    app.state.runtime_template_resolver = None
